File: src/ExprTree.java
/**
 * Created by Ivan on 5/27/2018.
 */
public class ExprTree {
    ExprNode root;
    ExprNode current;

    /**
     * Creates an empty tree
     */
    public ExprTree() {
        current = root = null;
    }

    /**
     * Creates a tree with 1 node
     * @param root root of the tree
     */
    public ExprTree(ExprNode root) {
        current = this.root = root;
    }

    //public ExprTree(ExprTree source) {}

    /**
     * Adopts the current working node
     * @param parent node's new parent
     * @return true on success
     */
    public boolean adoptCurrent(ExprNode parent) {
        return false;
    }

    /**
     * Retrieves the current working node
     * @return pointer to the node
     */
    public ExprNode getCurrent() {
        return null;
    }

    /**
     * Adds a new node at the current working location
     * @param node node to be added
     * @return true on success
     */
    public boolean add(ExprNode node) {
        return false;
    }



    static class ExprNode {
        enum Type {VAL, ADD, SUB, MUL, DIV}

        ExprNode.Type type;
        ExprNode[] children;

        public ExprNode(ExprNode.Type type) {
            this.type = type;
            this.children = new ExprNode[2];
        }

        public ExprNode(ExprNode.Type type, ExprNode ... children) {
            this.type = type;
            this.children = children;
        }

        static int sizeChildren(ExprNode.Type type) {
            switch (type) {
                case VAL: return 1;
                default: return 2;
            }
        }
    }
}

File: src/ExprTree.cpp
#include "ExprTree.h"

ExprNode::ExprNode(ExprType t) {
    left = right = NULL;
    type = t;
    value = NULL;
}

/**
 * Constructs a value node right away from the argument provided
 */
ExprNode::ExprNode(int v) {
    left = right = NULL;
    type = ExprType::VAL;
    value = new int;
    *value = v;
}

bool ExprNode::isHigher(const ExprNode * currentParent) const {
		if (type == ExprType::VAL) return false; // anything is higher than a value node
		else if ((type == ExprType::ADD || type == ExprType::SUB) &&
		(currentParent->type == ExprType::MUL || currentParent->type == ExprType::DIV))
			return false;
		else return true;
}

// ============ END NODE. BEGIN TREE ===========================

ExprTree::ExprTree() {
    currentParent = current = root = NULL;
}

int ExprTree::add(ExprNode * node) {
    if (!root) {
        currentParent = current = root = node;
    }
    else {
		if (node->isHigher(currentParent)) {

		}
		else {

		}
    }
}
